src/plugins/cache/expirationPlugin.ts

```diff
@@ -0,0 +1,125 @@
+import { logger } from "../../core/logger";
+import { CacheQueryMatchOptions } from "../../strategy/types";
+import { StrategyPlugin } from "../interfaces/strategyPlugin";
+
+class ExpirationPlugin implements StrategyPlugin {
+  private readonly maxEntries: number;
+  private readonly maxAgeSeconds: number;
+
+  constructor({
+    maxEntries,
+    maxAgeSeconds,
+  }: { maxEntries?: number; maxAgeSeconds?: number } = {}) {
+    this.maxAgeSeconds = maxAgeSeconds || 30 * 24 * 3_600;
+    this.maxEntries = maxEntries || Infinity;
+  }
+
+  async cachedResponseWillBeUsed(options: {
+    cacheName: string;
+    request: Request;
+    matchOptions: CacheQueryMatchOptions;
+    cachedResponse: Response;
+    event?: ExtendableEvent | undefined;
+  }): Promise<Response | null> {
+    const now = Date.now();
+
+    const expirationDate = options.cachedResponse.headers.get("X-Expires");
+
+    if (expirationDate) {
+      const elapsedTime = new Date(expirationDate).getTime() - now;
+
+      if (elapsedTime < 0) {
+        const cache = await caches.open(options.cacheName);
+        await cache.delete(options.request, options.matchOptions);
+        console.log("cacheResponseWillBeUsed", options.request.url);
+        return options.cachedResponse;
+      }
+
+      options.cachedResponse.headers.set(
+        "X-Access-Time",
+        new Date(now).toUTCString()
+      );
+
+      return options.cachedResponse;
+    } else {
+      options.cachedResponse.headers.set(
+        "X-Access-Time",
+        new Date(now).toUTCString()
+      );
+
+      return options.cachedResponse;
+    }
+  }
+
+  async cacheWillUpdate(options: {
+    response: Response;
+    request: Request;
+    event?: ExtendableEvent | undefined;
+  }): Promise<Response | null> {
+    const now = Date.now();
+    console.log("cacheWillUpdate", options.request.url);
+
+    let newResponse = options.response.clone();
+    newResponse.headers.set(
```

Comment:
Yes, 👍. This should work, and I think I have an idea of why it was immutable.

In short, because of in JS, arguments are passed as values, not references. So, any direct modification in the method didn't update the response because we passed just the value of the response, and not the actual response itself. My theory